### ISSUE TITLE
libyang2: conflict between valgrind and coverage

### ISSUE THREAD

**Issue by Alanscut** [CESNET/libyang#832]
When compiling libyang2, `cmake` with `ENABLE_VALGRIND_TESTS=ON` and `ENABLE_COVERAGE=ON`, the valgrind test and the coverage statistics will fail.
```
$ cmake  -DENABLE_COVERAGE=ON ..
```
```
$ make test
```

![image](https://user-images.githubusercontent.com/21104054/62126526-6eb25c00-b302-11e9-8ab6-8f21db45636e.png)
```
$ make coverage
```
![image](https://user-images.githubusercontent.com/21104054/62126558-87bb0d00-b302-11e9-9b98-109b62446385.png)

so if we want to make coverage, we should turn off the ENABLE_VALGRIND_TESTS at first.
```
$ cmake -DENABLE_VALGRIND_TESTS=OFF  -DENABLE_COVERAGE=ON ..
```

**Comment by rkrejci** [comment]
I'm running valgrind tests just fine even with coverage on. Could you please share the output of the valgind tests? And also the log (not the images, text file) of the cmake?

**Comment by Alanscut** [comment]
hi rkrejci, cmake is ok. when I run `make test`, the testcase with valgrind all fails:
```
make test
Running tests...
Test project /data/libyang2/build
      Start  1: src_set
 1/28 Test  #1: src_set ............................   Passed    0.00 sec
      Start  2: src_common
 2/28 Test  #2: src_common .........................   Passed    0.00 sec
      Start  3: src_context
 3/28 Test  #3: src_context ........................   Passed    0.02 sec
      Start  4: src_hash_table
 4/28 Test  #4: src_hash_table .....................   Passed    0.01 sec
      Start  5: src_xml
 5/28 Test  #5: src_xml ............................   Passed    0.00 sec
      Start  6: src_parser_yang
 6/28 Test  #6: src_parser_yang ....................   Passed    0.07 sec
      Start  7: src_tree_schema
 7/28 Test  #7: src_tree_schema ....................   Passed    0.01 sec
      Start  8: src_tree_schema_compile
 8/28 Test  #8: src_tree_schema_compile ............   Passed    0.10 sec
      Start  9: src_tree_schema_helpers
 9/28 Test  #9: src_tree_schema_helpers ............   Passed    0.01 sec
      Start 10: src_printer_yang
10/28 Test #10: src_printer_yang ...................   Passed    0.01 sec
      Start 11: src_tree_data
11/28 Test #11: src_tree_data ......................   Passed    0.01 sec
      Start 12: src_parser_xml
12/28 Test #12: src_parser_xml .....................   Passed    0.02 sec
      Start 13: src_printer_xml
13/28 Test #13: src_printer_xml ....................   Passed    0.01 sec
      Start 14: features_types
14/28 Test #14: features_types .....................   Passed    0.05 sec
      Start 15: src_set_valgrind
15/28 Test #15: src_set_valgrind ...................***Failed    0.42 sec
      Start 16: src_common_valgrind
16/28 Test #16: src_common_valgrind ................***Failed    0.42 sec
      Start 17: src_context_valgrind
17/28 Test #17: src_context_valgrind ...............***Failed    1.14 sec
      Start 18: src_hash_table_valgrind
18/28 Test #18: src_hash_table_valgrind ............***Failed    0.89 sec
      Start 19: src_xml_valgrind
19/28 Test #19: src_xml_valgrind ...................***Failed    0.46 sec
      Start 20: src_parser_yang_valgrind
20/28 Test #20: src_parser_yang_valgrind ...........***Failed    2.62 sec
      Start 21: src_tree_schema_valgrind
21/28 Test #21: src_tree_schema_valgrind ...........***Failed    0.88 sec
      Start 22: src_tree_schema_compile_valgrind
22/28 Test #22: src_tree_schema_compile_valgrind ...***Failed    3.74 sec
      Start 23: src_tree_schema_helpers_valgrind
23/28 Test #23: src_tree_schema_helpers_valgrind ...***Failed    0.85 sec
      Start 24: src_printer_yang_valgrind
24/28 Test #24: src_printer_yang_valgrind ..........***Failed    0.86 sec
      Start 25: src_tree_data_valgrind
25/28 Test #25: src_tree_data_valgrind .............***Failed    0.95 sec
      Start 26: src_parser_xml_valgrind
26/28 Test #26: src_parser_xml_valgrind ............***Failed    1.07 sec
      Start 27: src_printer_xml_valgrind
27/28 Test #27: src_printer_xml_valgrind ...........***Failed    1.12 sec
      Start 28: features_types_valgrind
28/28 Test #28: features_types_valgrind ............***Failed    2.09 sec

50% tests passed, 14 tests failed out of 28

Total Test time (real) =  17.83 sec

The following tests FAILED:
         15 - src_set_valgrind (Failed)
         16 - src_common_valgrind (Failed)
         17 - src_context_valgrind (Failed)
         18 - src_hash_table_valgrind (Failed)
         19 - src_xml_valgrind (Failed)
         20 - src_parser_yang_valgrind (Failed)
         21 - src_tree_schema_valgrind (Failed)
         22 - src_tree_schema_compile_valgrind (Failed)
         23 - src_tree_schema_helpers_valgrind (Failed)
         24 - src_printer_yang_valgrind (Failed)
         25 - src_tree_data_valgrind (Failed)
         26 - src_parser_xml_valgrind (Failed)
         27 - src_printer_xml_valgrind (Failed)
         28 - features_types_valgrind (Failed)
Errors while running CTest
Makefile:118: recipe for target 'test' failed
make: *** [test] Error 8
```
I chose one of the testcases(`src_tree_schema`) to test with valgrind:
```
root:/data/libyang2/build valgrind --tool=memcheck --leak-check=yes   --show-reachable=yes ./tests/src_tree_schema
==27559== Memcheck, a memory error detector
==27559== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==27559== Using Valgrind-3.12.0 and LibVEX; rerun with -h for copyright info
==27559== Command: ./tests/src_tree_schema
==27559==
[==========] Running 1 test(s).
[ RUN      ] test_getnext
[       OK ] test_getnext
[==========] 1 test(s) run.
[  PASSED  ] 1 test(s).
==27559== Syscall param write(buf) points to uninitialised byte(s)
==27559==    at 0x5254CA0: __write_nocancel (in /lib64/libpthread-2.22.so)
==27559==    by 0x4E38815: SendHookMessage (in /lib64/libachk.so)
==27559==    by 0x4E38D54: checkFileInfo (in /lib64/libachk.so)
==27559==    by 0x4E38061: acc_achkon_open (in /lib64/libachk.so)
==27559==    by 0x4E376E0: open (in /lib64/libachk.so)
==27559==    by 0x4AA80A: __gcov_open (gcov-io.c:101)
==27559==    by 0x4AB3BD: gcov_exit (libgcov.c:523)
==27559==    by 0x59E7138: __run_exit_handlers (in /lib64/libc-2.22.so)
==27559==    by 0x59E7184: exit (in /lib64/libc-2.22.so)
==27559==    by 0x59D06EB: (below main) (in /lib64/libc-2.22.so)
==27559==  Address 0xffeffa2d2 is on thread 1's stack
==27559==  in frame #2, created by checkFileInfo (???:)
==27559==
==27559==
==27559== HEAP SUMMARY:
==27559==     in use at exit: 167 bytes in 2 blocks
==27559==   total heap usage: 3,275 allocs, 3,273 frees, 2,036,925 bytes allocated
==27559==
==27559== 32 bytes in 1 blocks are still reachable in loss record 1 of 2
==27559==    at 0x4C2B1B0: calloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==27559==    by 0x5D5450F: _dlerror_run (in /lib64/libdl-2.22.so)
==27559==    by 0x5D53F90: dlopen@@GLIBC_2.2.5 (in /lib64/libdl-2.22.so)
==27559==    by 0x4E37CB0: loadRealFunc (in /lib64/libachk.so)
==27559==    by 0x4E37320: hookSyscheckInit (in /lib64/libachk.so)
==27559==    by 0x4E3766D: open (in /lib64/libachk.so)
==27559==    by 0x4AA80A: __gcov_open (gcov-io.c:101)
==27559==    by 0x4AB3BD: gcov_exit (libgcov.c:523)
==27559==    by 0x59E7138: __run_exit_handlers (in /lib64/libc-2.22.so)
==27559==    by 0x59E7184: exit (in /lib64/libc-2.22.so)
==27559==    by 0x59D06EB: (below main) (in /lib64/libc-2.22.so)
==27559==
==27559== 135 bytes in 1 blocks are still reachable in loss record 2 of 2
==27559==    at 0x4C29080: malloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==27559==    by 0x4E390DE: initNoHookProcessList (in /lib64/libachk.so)
==27559==    by 0x4E37274: setHookFlag (in /lib64/libachk.so)
==27559==    by 0x4E37386: hookSyscheckInit (in /lib64/libachk.so)
==27559==    by 0x4E3766D: open (in /lib64/libachk.so)
==27559==    by 0x4AA80A: __gcov_open (gcov-io.c:101)
==27559==    by 0x4AB3BD: gcov_exit (libgcov.c:523)
==27559==    by 0x59E7138: __run_exit_handlers (in /lib64/libc-2.22.so)
==27559==    by 0x59E7184: exit (in /lib64/libc-2.22.so)
==27559==    by 0x59D06EB: (below main) (in /lib64/libc-2.22.so)
==27559==
==27559== LEAK SUMMARY:
==27559==    definitely lost: 0 bytes in 0 blocks
==27559==    indirectly lost: 0 bytes in 0 blocks
==27559==      possibly lost: 0 bytes in 0 blocks
==27559==    still reachable: 167 bytes in 2 blocks
==27559==         suppressed: 0 bytes in 0 blocks
==27559==
==27559== For counts of detected and suppressed errors, rerun with: -v
==27559== Use --track-origins=yes to see where uninitialised values come from
==27559== ERROR SUMMARY: 26 errors from 1 contexts (suppressed: 0 from 0)

```

**Comment by rkrejci** [comment]
Please, could you share the output of the cmake? coverage works only with gcc, so the source of the problem here can be the compiler.

**Comment by Alanscut** [comment]
do you mean the CMakeOutput.log under the /build/CMakeFiles directory?

[CMakeOutput.log](https://github.com/CESNET/libyang/files/3446714/CMakeOutput.log)

**Comment by rkrejci** [comment]
That's fine, thanks. It seems that there were some issues with gcov + GCC 4.8 you are using. They should be fixed since GCC 4.9, so upgrading should help.

btw, coverage is supposed to be just a helper functionality which shouldn't be necessary/helpful for the most of users (it is good only for the developers and testers). Are you sure that you want/need it?

**Comment by Alanscut** [comment]
ok, I will upgrade my gcc. 
Most users really don't care about coverage, but I am interested in libyang2's source code in addition to using libyang. By learning the refactoring of libyang2, I hope to learn more coding techniques. As a developer, I really care about the coverage of a piece of software, and comprehensive testcases make it easier to spot potential problems.
